Merge branch Feature/aggregate statistics generator into develop

AdamStanger opened the pull request:
Files added:

- frontend/src/components/StatisticsPage.jsx: Page containing the tabe for all the projectile points in the catalogue
- frontend/src/components/StatisticsModal.jsx: Modal for displaying charts for the generated statistics
- frontend/tests/statisticsPage.unit.test.jsx

Modifications:

- Functions in backend/controllers/aggregateStatisticsGenerator.js changed to be async to fix issue when passing data to the frontend
- Unit tests for those functions updated based on refactor
- Path to statistics page added (localhost:8080/stats)

brj496 committed:
Created aggregateStatisticsController.js and aggregateStatisticsGenerator.test.js, which include stubbed functions for generating the aggregate statistics, and the unit tests for those functions

brj496 committed:
Fixed some minor issues with my test suites, coded functionality for the materialPercentages() function.

brj496 committed:
implemented average dimensions (NOTE: this may need to change based on how the dimensions are stored into the database.)

brj496 committed:
implemented projectilePointPercentage() and adjusted its tests based on new understanding.

brj496 committed:
made changes to theses routes, and create helper functions so that the database calls may be made internally.

brj496 committed:
added some loggers to help debug issues.

brj496 committed:
implemented code for aggregateSiteStatisitics()

brj496 committed:
Adding Package-lock and Package

brj496 committed:
Cleaning up some TODOs and removing some logger.info

brj496 committed:
implements functionality for aggregateCatalogueStatistics() and aggregatePointTypeStatistics(), and moves endpoint call for artifactTypes into functions for easier calling/testing.

brj496 committed:
forgot to include the helperFile for artifactTypes endpoint in previous commit.

brj496 committed:
Creates routes for aggregateStatisticsGenerators and slightly adjusts outputs for the statistics generots to output stings based on problem encountered.

Commit:
Created integration tests for testing functions on empty catalogue

brj496 committed:
applies some bug fixes, and fixes broken test for the aggregateStatisticsGenerators
also fixes the logger require on like 7 in the aggregateStatisticsController

brj496 committed:
cleans up some left over comments/TODOs and commented code.

brj496 committed:
removes package and package-lock from root

Commit:
Merge branch 'develop' of https://github.com/UniversityOfSaskatchewanCMPT371/term-project-2024-team-4 into feature/aggregate-statistics-generator

Commit:
Created unit tests for aggregate statistics page

Commit:
Made base for statistics page, added route as '/stats' in App.jsx., fixed import issue in statisticsPage.unit.test.js

Commit:
Fixed issues preventing unit tests from running; Removed unit tests that weren't needed (some had already been tested in other components)

lecegues committed:
Updated `users.js` to extract JWT_SECRET from environment variable

lecegues committed:
Updated `docker-compose.yaml` to use .env variables

- especially to account for private but constant JWToken

lecegues committed:
Temporary change to workflow file to accomodate for `JWT_SECRET` .env variable

- Will be changed in a later patch to be private

Commit:
Created base for statistics modal. Appears when 'Generate Statistics' button is pressed; has areas for material and point statistics; Charts need to be created and added.

lecegues committed:
Removed POSTGRES_PASSWORD env variable from docker-compose file

- POSTGRES_PASSWORD could not reference .env file `DB_PASS` variable
- Therefore, its better to add POSTGRES_PASSWORD in .env file

lecegues committed:
Created unit tests for authentication feature

lecegues committed:
Created authentication middleware to authenticate admin-only actions

lecegues committed:
Removed unnecessary quotes that are causing parsing errors

Commit:
Pushing current code. Bug with MUI datagrid.

Commit:
Fixed issue from previous commit. Removed basic table, replaced with datagrid; axios fetch properly gets projectile points; point data is parsed into rows for data table.

Commit:
Fixed issues with aggregateStatisticsGenerator.js; Added relations in catalogueHelper.js to make statistics easier; Pie Charts need to be fixed in statisticsModal.jsx

Commit:
Pie chart now renders data given by aggregate statistics generator, some code clean up, added some functionality to aggregateStatisticsGenerators.js for converting maps to objects for passing

lecegues committed:
Merge branch 'patch/issue-92' into feature/aggregate-statistics-generator

Commit:
Fixing issue with unit test involving opening modal not being able to run, added window.matchMedia to setup.js. Removed redundant test

lecegues committed:
Merge branch 'feature/aggregate-statistics-generator' of https://github.com/UniversityOfSaskatchewanCMPT371/term-project-2024-team-4 into feature/aggregate-statistics-generator

Commit:
Fixing aggregateSiteStatistics() and aggregatePointTypeStatistics() in aggregateStatisticsController.js so they are async (needed for when called from the frontend); Fixed unit tests after changing the functions so they all pass again; slight refactoring of StatisticsPage.jsx and StatisticsModal.jsx

Commit:
Fixing some issues causing the linter to fail in pull request #109

brj496 committed:
Adds integration, end to end, and smoke tests for the aggregateStatisticsGenerator feature. (note the statisitcsPage.spec.js e2e tests are in the smoke test folder as thats the only place they would function)

brj496 committed:
fixes issue with test failing, the close button wasn't clicked and could not hide the modal

brj496 committed:
moves statistics page e2e tests from the smoke test folder to the testing folder. so that they do not run during building in the pipeline

brj496 committed:
creates helperFiles to use as test hooks for the creation and manipulation of data without the use of endpoints. Also applies jewel's refactoring of the endpoints.

also fixes issues with relations in database for sites and artifactType

Creates the following hooks:
baseShapesHelper.js
bladeShapesHelper.js
crossSectionsHelper.js
culturesHelper.js
haftingShapesHelper.js
materialsHelper.js
periodsHelper.js
regionsHelper.js

Updates the following helperFiles due to Jewel's refactor
artifactsHelper.js
artifactTypesHelper.js
cataloguesHelper.js

brj496 committed:
Fixes feature 4 integration tests so that they create/delete the data for the tests while they are running

also fixes issue with front end package-lock being out of sync

brj496 committed:
adds express-validator package to backend

brj496 committed:
adds puppeteer to front end packages

lecegues committed:
Added longer timeouts for integration tests that could take a long amount of time

lecegues committed:
Upped the timeout for long-running tests and changed test reporter to `verbose` to see potential errors in tests

lecegues commented on backend/config/authenticate.js:
This file should be deleted (remnants from the merge from develop). This file was moved to the `middleware` folder

brj496 replied:
Removed authenticate.js from backend config

lecegues commented on backend/package.json:
This should **not** get rid of these scripts and dependencies. Could be a mistake from the merge from develop.

brj496 replied:
Made sure to add back removed dependencies

lecegues commented on frontend/package.json:
@vitest/coverage-v8 package should **not** be touched. Again, probably something missed due to the merge.

lecegues commented on frontend/src/components/Home.jsx:
Was box supposed to be first or not? If I remember, I know that the order of dependencies can sometimes cause MUI to crash... Please verify

brj496 replied:
I'm not sure, I don't remember messing with this file, so it may have been changed due to the merge

brj496 replied:
turns out they are in the right order

lecegues commented on frontend/tests/setup.js:
As long as this addition to `setup.js` does not interfere with other tests, then it is ok. Just noting this because **every test** will be referencing this `setup.js` file.

brj496 replied:
I think Dani is the one who made this file, but all the tests function on my machine.

lecegues reviewed (changes requested):
Please view my comments for some changes. 
Additionally, am I correct in assuming that all of the `routes` were changed because they now use the corresponding `helperFile`? Or were the changes to the `route` files accidental from the merge?

brj496 commented:
> Please view my comments for some changes. Additionally, am I correct in assuming that all of the `routes` were changed because they now use the corresponding `helperFile`? Or were the changes to the `route` files accidental from the merge?

each route was changed to use its corresponding helperFile. They are correct as they are right now

brj496 committed:
removes authenticate.js file, fixes issues with removed dependencies, added back scripts that were removed from package files, also swaps Box import from mui to be first.

brj496 committed:
fixes desync with package-lock and packages

brj496 committed:
fixes issue with MUI dependencies being in the wrong issue

lecegues commented on testing/smoke-test/smoke.spec.js:
You've accidentally added back the smoke tests that were removed, and accidentally removed the new smoke test(edge cases). Please change this

brj496 replied:
added back new tests and removed old tests

brj496 committed:
replaces old smoke tests with new ones from develop

lecegues reviewed (approved):
Reviewed changes were fixed.

lecegues committed:
Merge branch 'develop' into feature/aggregate-statistics-generator

lecegues committed:
Updated package-lock.json after module updates

lecegues commented on backend/routes/artifacts.js:
```diff
@@ -78,19 +30,11 @@ router.post("/", async (req, res) => {
-	} catch (error) {
-		logger.error("Error fetching Artifacts: " + error.message);
-		res.status(500).json({ error: error.message });
+	const response = await artifactsHelper.newArtifact(req);
```
GET request, but calling a POST helper

brj496 replied:
Fixed Issue, replaced with correct function call

lecegues commented on backend/routes/artifacts.js:
```diff
@@ -103,20 +47,11 @@ router.get("/", async (req, res) => {
-	} catch (error) {
-		logger.error(`Error fetching Artifact with ID ${id}: ${error.message}`);
-		res.status(500).json({ error: error.message });
+	const response = await artifactsHelper.newArtifact(req);
```
GET request, calling POST helper

brj496 replied:
Fixed Issue, replaced with correct function call

lecegues commented on backend/routes/artifacts.js:
```diff
@@ -129,48 +64,11 @@ router.get("/:id", async (req, res) => {
-	} catch (error) {
-		logger.error(`Error updating Artifact with ID ${id}: ${error.message}`);
-		res.status(500).json({ error: error.message });
+	const response = await artifactsHelper.newArtifact(req);
```
PUT req, calling a POST helper

brj496 replied:
Fixed Issue, replaced with correct function call

lecegues commented on backend/routes/artifacts.js:
```diff
@@ -183,18 +81,11 @@ router.put("/:id", async (req, res) => {
-	} catch (error) {
-		logger.error(`Error deleting Artifact with ID ${id}: ${error.message}`);
-		res.status(500).json({ error: error.message });
+	const response = await artifactsHelper.newArtifact(req);
```
DELETE req, using a POST helper

brj496 replied:
Fixed Issue, replaced with correct function call

lecegues commented on backend/routes/crossSections.js:
```diff
@@ -124,18 +92,11 @@ router.put("/:id", async (req, res) => {
-	} catch (error) {
-		logger.error(`Error deleting CrossSection with ID ${id}: ${error.message}`);
-		res.status(500).json({ error: error.message });
+	const response = await crossSectionsHelper.updateCrossSection(req);
```
DELETE req, using UPDATE helper

brj496 replied:
Fixed Issue, replaced with correct function call

lecegues commented on backend/routes/materials.js:
```diff
@@ -72,22 +50,14 @@ router.get("/", async (req, res) => {
-	} catch (error) {
-		logger.error("Error fetching Material:", error);
-		res.json({ error: error.message });
+	const response = await materialsHelper.newMaterial(req);
```
GET req, calling a POST helper

brj496 replied:
Fixed Issue, replaced with correct function call

lecegues commented on backend/routes/materials.js:
```diff
@@ -103,35 +73,17 @@ router.get("/:id", async (req, res) => {
-	} catch (error) {
-		logger.error("Error updating Material:", error);
-		res.json({ error: error.message });
+	const response = await materialsHelper.newMaterial(req);
```
UPDATE req, calling a POST helper

brj496 replied:
Fixed Issue, replaced with correct function call

lecegues commented on backend/routes/materials.js:
```diff
@@ -145,19 +97,14 @@ router.put("/:id", async (req, res) => {
-	} catch (error) {
-		logger.error("Error deleting Material:", error);
-		res.json({ error: error.message });
+	const response = await materialsHelper.newMaterial(req);
```
DELETE req, calling a POST helper

brj496 replied:
Fixed Issue, replaced with correct function call

lecegues commented on backend/routes/periods.js:
```diff
@@ -116,20 +79,14 @@ router.put("/:id", async (req, res) => {
-	} catch (error) {
-		logger.error("Error deleting Period:", error);
-		res.json({ error: error.message });
+	const response = await periodsHelper.updatePeriod(req);
```
DELETE req, calling UPDATE helper

brj496 replied:
Fixed Issue, replaced with correct function call

lecegues commented on backend/routes/regions.js:
```diff
@@ -34,16 +21,11 @@ router.get("/", async (req, res) => {
-	} catch (error) {
-		logger.error("Error creating new Region:", error);
-		res.json({ error: error.message });
+	const response = await regionsHelper.getAllRegions();
```
POST req, calling a GET helper

brj496 replied:
Fixed Issue, replaced with correct function call

lecegues commented on backend/routes/regions.js:
```diff
@@ -56,25 +38,14 @@ router.post("/", async (req, res) => {
-	} catch (error) {
-		logger.error("Error fetching Region:", error);
-		res.json({ error: error.message });
+	const response = await regionsHelper.getAllRegions();
```
SINGLE GET req, calling a GET ALL helper

brj496 replied:
Fixed Issue, replaced with correct function call

lecegues commented on backend/routes/regions.js:
```diff
@@ -87,25 +58,14 @@ router.get("/:id", async (req, res) => {
-	} catch (error) {
-		logger.error("Error updating Region:", error);
-		res.json({ error: error.message });
+	const response = await regionsHelper.getAllRegions();
```
UPDATE req, calling a GET helper

brj496 replied:
Fixed Issue, replaced with correct function call

lecegues commented on backend/routes/regions.js:
```diff
@@ -118,20 +78,14 @@ router.put("/:id", async (req, res) => {
-	} catch (error) {
-		logger.error("Error deleting Region:", error);
-		res.json({ error: error.message });
+	const response = await regionsHelper.getAllRegions();
```
DELETE req, calling a GET helper

brj496 replied:
Fixed Issue, replaced with correct function call

lecegues reviewed (changes requested):
There are some API requests that call different helpers than what is intended. I have outlined which ones need ot be changed

brj496 committed:
Fixes issues with routes not using the correct helperFile calls

brj496 committed:
Merge branch 'feature/aggregate-statistics-generator' of https://github.com/UniversityOfSaskatchewanCMPT371/term-project-2024-team-4 into feature/aggregate-statistics-generator

brj496 committed:
removes unused requires from haftingShapes.js

brj496 committed:
Fixes issue with statisticsgenerators not returning the correct string when trying to catch a non-existent site.

brj496 committed:
fixes issue with statisticsPage.unit.test.js failing due to not using memoryRouter

lecegues committed:
Updated frontend request URLs & fix to #115

- Additional bug encountered relating to #115 (Login feature) and fixed
    - Fixed by allowing cross-site requests
- Updated frontend request URLs to reference global axios configuration

lecegues committed:
Merge branch 'feature/aggregate-statistics-generator' of https://github.com/UniversityOfSaskatchewanCMPT371/term-project-2024-team-4 into feature/aggregate-statistics-generator

lecegues committed:
Added missing file from previous commit (c02cb2e71b9ba52218179199ea0f69ad397b9f9a)

lecegues committed:
Added return statements to all backend res (responses) to fix multiple responses

- Multiple responses were being sent because return statemnts were not specified
- Additionally, it caused some tests to fail

lecegues reviewed (approved):
Reviewed changes were fixed. Additionally, some tests stopped working which revealed some bugs (#115) and were fixed. Also removed failing `Login.integ.test.jsx`; refer to issue #116. Branch should be good for merge.

funmise reviewed (approved):
Helper files are a good change to the routes
